src/Microsoft.Health.Fhir.CosmosDb.UnitTests/Features/Storage/CosmosFhirDataStoreTests.cs

```diff
@@ -0,0 +1,204 @@
+            ICosmosQuery<int> cosmosQuery = Substitute.For<ICosmosQuery<int>>();
+            _cosmosQueryFactory.Create<int>(default, default).ReturnsForAnyArgs(cosmosQuery);
+
+            FeedResponse<int> response = CreateFeedResponse(0, 2, "token");
+
+            cosmosQuery.ExecuteNextAsync().ReturnsForAnyArgs(ci => response, ci => throw new RequestRateExceededException(null));
+            cosmosQuery.HasMoreResults.Returns(true);
+
+            (IReadOnlyList<int> results, string continuationToken) = await _dataStore.ExecuteDocumentQueryAsync<int>(
+                new QueryDefinition("abc"),
+                new QueryRequestOptions { MaxItemCount = 10 });
+
+            Assert.Equal(Enumerable.Range(0, 2), results);
+            Assert.Equal("token", continuationToken);
+        }
+
+        [Fact]
+        public async Task GivenAQuery_WhenFetchingSubsequentPagesTimesOut_ReturnsExistingResults()
+        {
+            ICosmosQuery<int> cosmosQuery = Substitute.For<ICosmosQuery<int>>();
+            _cosmosQueryFactory.Create<int>(default, default).ReturnsForAnyArgs(cosmosQuery);
+
+            FeedResponse<int> response = CreateFeedResponse(0, 2, "token");
+
+            cosmosQuery.ExecuteNextAsync().ReturnsForAnyArgs(ci => response, ci => throw new OperationCanceledException());
+            cosmosQuery.HasMoreResults.Returns(true);
+
+            var time = DateTimeOffset.UtcNow;
+            _cosmosDataStoreConfiguration.SearchEnumerationTimeoutInSeconds = 0;
+
+            // lock the time
+            using (Mock.Property(() => ClockResolver.UtcNowFunc, () => time))
+            {
+                (IReadOnlyList<int> results, string continuationToken) =
+                    await _dataStore.ExecuteDocumentQueryAsync<int>(
+                        new QueryDefinition("abc"),
+                        new QueryRequestOptions { MaxItemCount = 10 });
+
+                Assert.Equal(Enumerable.Range(0, 2), results);
+                Assert.Equal("token", continuationToken);
+            }
+        }
+
+        [Fact]
+        public async Task GivenAQueryWhereItemCountCanBeExceeded_WhenExecuted_FetchesSubsequentPages()
+        {
+            CreateResponses(
+                10,
+                null,
+                CreateFeedResponse(0, 0, "1"),
+                CreateFeedResponse(0, 1, "2"),
+                CreateFeedResponse(0, 0, "3"),
+                CreateFeedResponse(1, 1, null));
+
+            (IReadOnlyList<int> results, string continuationToken) = await _dataStore.ExecuteDocumentQueryAsync<int>(
+                new QueryDefinition("abc"),
+                new QueryRequestOptions { MaxItemCount = 10 },
+                mustNotExceedMaxItemCount: false);
+
+            Assert.Equal(Enumerable.Range(0, 2), results);
+            Assert.Null(continuationToken);
+        }
+
+        [Fact]
+        public async Task GivenAQueryWhereItemCountMustNotBeExceeded_WhenExecuted_FetchesSubsequentPagesByIssuingNewQueries()
```

Comment:
I don't think this test shows the desired behavior. The second half of the first CreateResponses is the same as the second CreateResponses. It looks like this test would pass whether it was calling the new query or the old one.

Reply:
Good catch. I've updated the test.